BUG: SparseSeries init from dict fixes

kernc opened the pull request:
- [X] closes #16905
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [X] whatsnew entry

kernc committed:
BUG: SparseSeries init from dict fixes

kernc commented:
Tests copied/adapted from `tests.series.test_constructors`.

jreback commented on pandas/tests/sparse/test_series.py:
```diff
@@ -1361,3 +1364,96 @@ def test_numpy_func_call(self):
+
+    data = A(('col%s' % i, np.random.random()) for i in range(12))
+    s = SparseSeries(data)
+    tm.assert_numpy_array_equal(s.values.values, np.array(list(data.values())))
```
can you use assert_sp_series_equal (you can pass check_list=False) and then add an assert about the column ordering

jreback commented on pandas/tests/sparse/test_series.py:
```diff
@@ -1361,3 +1364,96 @@ def test_numpy_func_call(self):
+        return dict(zip((constructor(x) for x in dates_as_str), values))
+
+    data_datetime64 = create_data(np.datetime64)
+    data_datetime = create_data(lambda x: datetime.strptime(x, '%Y-%m-%d'))
```
can you parameterize this test

jreback commented on pandas/tests/sparse/test_series.py:
```diff
@@ -1361,3 +1364,96 @@ def test_numpy_func_call(self):
+    expected = SparseSeries([x[1] for x in _d],
+                            index=pd.Index([x[0] for x in _d],
+                                           tupleize_cols=False))
+    ser = ser.reindex(index=expected.index)
```
use result=

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -179,6 +179,7 @@ Sparse
 ^^^^^^
 
 
+- Bug in instantiating :class:`SparseSeries` from ``dict`` with or without ``index`` (:issue:`16905`)
```
``index=`` kwarg

kernc committed:
fixup! BUG: SparseSeries init from dict fixes

jreback commented on pandas/tests/sparse/test_series.py:
```diff
@@ -1361,3 +1364,89 @@ def test_numpy_func_call(self):
+
+
+def test_constructor_dict():
+    d = {'a': 0., 'b': 1., 'c': 2.}
```
you might be able to move some of these into ``from pandas.tests.series.test_api import SharedWithSparse`` whech we already import (rather than directly copying them).

gfyoung commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -179,6 +179,7 @@ Sparse
 ^^^^^^
 
 
+- Bug in instantiating :class:`SparseSeries` from ``dict`` with or without ``index=`` kwarg (:issue:`16905`)
```
If it doesn't matter whether `index` is passed in, why mention it in the description?

kernc replied:
It's a reference to the _two issues_ fixed. With, the result was invalid; without, it crashed.

gfyoung replied:
Fair enough, though ultimately instantiating from ``dict`` just didn't work at all though, which could be considered a single bug (also you're only referencing one issue here).  Note that without further context, people won't be aware of that difference, so it is preferable to be concise.

jreback commented:
@kernc lmk if you are able to share some test code with Series, can always do a followup.

kernc committed:
Move several tests to SharedWithSparse

pep8speaks commented:
Hello @kernc! Thanks for updating the PR.



 - In the file [`pandas/tests/series/test_api.py`](https://github.com/pandas-dev/pandas/blob/5dcec572dff255a855d2a20f7a78baba36a3cb6b/pandas/tests/series/test_api.py), following are the PEP8 issues :

> [Line 166:33](https://github.com/pandas-dev/pandas/blob/5dcec572dff255a855d2a20f7a78baba36a3cb6b/pandas/tests/series/test_api.py#L166): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

kernc committed:
update whatsnew

kernc commented:
Was able to share some test code with Series with as little as possible effort. Not sure if OK, though.

jreback commented on pandas/tests/series/test_api.py:
```diff
@@ -117,9 +118,78 @@ def test_to_sparse_pass_name(self):
+
+        result = self.Series(d, index=['b', 'c', 'd', 'a'])
+        expected = self.Series([1, 2, np.nan, 0], index=['b', 'c', 'd', 'a'])
+        tm.assert_series_equal(result, expected)
```
This doesn't check sparseness. I might modify assert_series_equal to dispatch to assert_sp_series_equal if *both* are SparseSeries.

gfyoung replied:
In light of my comment <a href="https://github.com/pandas-dev/pandas/pull/16906#issuecomment-315623547">below</a>, I'd rather that we keep sparse equality and `Series` equality checks separate.  Perhaps if we could write a function like:

~~~python
def _check_series_equal(self, left, right):
     ...
~~~

that dispatches to `tm.assert_series_equal` OR `tm.assert_sp_series_equal` depending on test class.

kernc committed:
assert_series_equal dispatch to sp_series_equal if both are sparse

gfyoung commented:
> Was able to share some test code with Series with as little as possible effort. Not sure if OK, though.

I'm a little hesitant about this code-sharing because the readability decreased IMO.  `self.Series` for me a just a little harder to understand.

kernc commented:
How else would you share code without `self.Series`? And if you'd rather not share code, how else would you ensure the two _is-a-Series_ types get roughly the same amount of use case and API coverage?

jreback commented:
@kernc this looks fine. ping on green.

kernc committed:
Fix a failing test ...

that was failing due to introduced dispatch to assert_sp_series_equal
being too strict.

gfyoung commented:
> How else would you share code without self.Series?

I personally find the name a little confusing because I only think the `Series` class and not `SparseSeries`, even though the latter is a subclass of the former. 😄 

A name like `self.series_klass` might have been easier because it doesn't have that confusion of naming.

kernc committed:
fixup! Fix a failing test ...

kernc committed:
self.Series -> self.series_klass

gfyoung commented on pandas/tests/series/test_api.py:
```diff
@@ -117,9 +118,78 @@ def test_to_sparse_pass_name(self):
+        d = {'a': 0., 'b': 1., 'c': 2.}
+        result = self.series_klass(d)
+        expected = self.series_klass(d, index=sorted(d.keys()))
+        tm.assert_series_equal(result, expected)
```
I think GitHub hid away my comment about this, but I think we should *not* interlace `assert_series_equal` and `assert_sp_series_equal`.  It reduces modularity, and that "assert_series_equal" is confusing for `SparseSeries` I find.

I propose that we do the following and define the following method:

~~~python
def assert_series_klass_equal(result, expected):
    klass_name = self.series_klass.__name__

    if klass_name == "Series":
        tm.assert_series_equal(result, expected)
    elif klass_name == "SparseSeries":
        tm.assert_sp_series_equal(result, expected)
    else:
        raise ValueError("Invalid 'series_klass' : {name}".format(name=klass_name))
~~~

That way you also don't need to modify `assert_series_equal`.

jreback commented:
@kernc would you be ok with merging #16960, close this, then you can refactor tests in a new PR?

kernc commented:
Of course.

kernc commented:
Continued in https://github.com/pandas-dev/pandas/pull/17050.
